File: 1/2020.10.13/433.Minimum Genetic Mutation/java/solution.java
class Solution {
    public int minMutation(String start, String end, String[] bank) {
        Set<String> bankSet = new HashSet<>();
        for (String s : bank) {
            bankSet.add(s);
        }
        if (!bankSet.contains(end)) {
            return -1;
        }
        char[] ca = {'A', 'C', 'G', 'T'};
        Deque<String> q = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();
        q.offerLast(start);
        visited.add(start);
        int level = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            level++;
            while (size-- > 0) {
                String word = q.pollFirst();
                char[] wordArray = word.toCharArray();
                for (int i = 0; i < wordArray.length; i++) {
                    char old = wordArray[i];
                    for (int j = 0; j < ca.length; j++) {
                        wordArray[i] = ca[j];
                        String newWord = new String(wordArray);
                        if (newWord.equals(end)) {
                            return level;
                        }
                        if (!visited.contains(newWord) && bankSet.contains(newWord)) {
                            q.offerLast(newWord);
                            visited.add(newWord);
                        }
                    }
                    wordArray[i] = old;
                }
            }
        }
        return -1;
    }
}

File: 1/2020.10.13/433.Minimum Genetic Mutation/python/solution.py
class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        if not end or not bank or end not in bank:
            return -1
        bank = set(bank)
        q = [(start, 0)]
        visited = {start}
        while q:
            word, level = q.pop(0)
            for i in range(len(word)):
                for c in ['A', 'C', 'G', 'T']:
                    intermediate_word = word[:i] + c + word[i+1:]
                    if intermediate_word == end:
                        return level + 1
                    if intermediate_word not in visited and intermediate_word in bank:
                        q.append((intermediate_word, level+1))
                        visited.add(intermediate_word)
        return -1
